Gutenberg: Add Goodreads Block

Aurorum opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes Automattic/wp-calypso#74494
Fixes #7920 
## Proposed changes:
<!--- Explain what functional changes your PR includes -->

Adds a block for Goodreads so that users can update to newer themes - it's currently only available as a legacy widget. This block is similar to the existing widget in that it relies on the default Goodreads styles, but I've added a little bit more customisation which Goodreads offers that was never included in the legacy widget. This can be seen in the Settings panel, and it just controls which books display.

Goodreads also offers a Grid widget that was never used in Jetpack. I feel like Gutenberg is perfect at handling multiple styles, so I've added it with this block.

| Grid Widget                                      | Default Widget                                     |
| -------------------------------------------- | -------------------------------------------- |
| ![Screenshot 1](https://github.com/Automattic/jetpack/assets/43215253/69c77d3e-90f2-4638-9735-64db0b774aec) | ![Screenshot 2](https://github.com/Automattic/jetpack/assets/43215253/6e297023-b2ef-4637-9b13-9fa2f7706528) |

I'm marking this as a draft because @supernovia mentioned in the original issue that it'd be good to support authors as well as readers. Unfortunately, I don't have an Authors Goodreads account, so I can't yet see how to implement those widgets. I'll do a bit more digging on that, but would be very grateful for any early feedback - it's been a while since I've tried my hand at making a block!

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No changes. Regarding the above section, I don't believe that I'm able to test my code on WordPress.com, and I'm not familiar with e2e tests for Jetpack.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

- You should be able to create a Goodreads block using a profile URL - so for example: https://www.goodreads.com/user/show/1176283-matt-mullenweg
- Verify that it loads properly and that each of the settings in the sidebar works 
- Hopefully nothing is too out of place! 

Notably, you should also be able to test this with an Author URL. The Goodreads Widget directs users to documentation on how to acquire their user ID because it's different for authors; with this block, it's much simpler and only requires your profile URL. Example author profile: https://www.goodreads.com/author/show/1406384

Aurorum committed:
Add slug

Aurorum committed:
First stab at Goodreads block

Aurorum committed:
Load block front-end

Aurorum commented on projects/plugins/jetpack/extensions/blocks/goodreads/edit.js:
```diff
@@ -0,0 +1,419 @@
+		}
+
+		const html = `
+		<style> [class^=gr_custom_container_] { border: 1px solid gray; border-radius: 10px; margin: auto; padding: 0 5px 10px 5px; background-color: #FFF; color: #000; width: 300px; }  [class^=gr_custom_header_] { border-bottom: 1px solid gray; width: 100%; padding: 10px 0; margin: auto; text-align: center; font-size: 120%; }  [class^=gr_custom_each_container_] { width: 100%; clear: both; margin: auto ...
```
There has got to be a better way of doing this...but no idea what that would be!

Aurorum committed:
Handle author profiles

Aurorum committed:
Endpoint for getting user ID

Aurorum committed:
Remove unneeded parameter

Aurorum commented:
Although it would still be nice to add the Author widgets, I can't seem to figure out how to get access to them right now, so I'm marking this as ready for review! Nevertheless, I've added some changes so that it's much simpler for authors to use this block anyway.

Whereas the Goodreads widget requires people to read documentation about how to acquire their User ID in case they give us their Author ID, now you can just input your profile URL and the block will handle distinguishing between the two.

Aurorum commented:
@jeherve I’m looking at adding tests for this block and I’d appreciate some advice - what does the `fixtures` folder in each blocks folder actually do? eg. https://github.com/Automattic/jetpack/tree/trunk/projects/plugins/jetpack/extensions/blocks/eventbrite/test/fixtures

Is this necessary for the tests? Also, apologies if I’ve just completely missed it, but is there any documentation which you could point me to on this? Thanks!!

supernovia commented:
Awesome, thanks @Aurorum !

jeherve commented:
@Aurorum The fixtures represent mock data for each block. That data is used in the `runBlockFixtureTests` function here, basically ensuring that the block structure outcome you have in the fixtures matches what's actually outputted when you add a block:
https://github.com/Automattic/jetpack/blob/260c45af538dab84b7ccdbad52b869c4c0911efb/projects/plugins/jetpack/extensions/shared/test/block-fixtures.js#L13-L20

You can run `cd projects/plugins/jetpack && pnpm test-extensions` to test this locally.

Aurorum committed:
Add tests

Aurorum commented:
Thanks @jeherve - I think this block is good to go, except for tests which I don't have much experience with. 

I'm a little stuck on how to get the block to display when embedded. I've drawn this from the Calendly Block tests, but it's still set as embedding the profile for me: 

```js
	test( 'display block', async () => {
		const attributes = {
		      ...defaultAttributes,
		      goodreadsId: '1176283',
		      userInput: 'https://www.goodreads.com/user/show/1176283-matt-mullenweg',
		    };
		render( <GoodreadsEdit { ...{ ...defaultProps, attributes } } /> );

		let iframe;
		await waitFor( () => ( iframe = screen.getByTitle( 'Goodreads' ) ) );

		expect( iframe ).toBeInTheDocument();
	} );
```

Other than that, I think this is ready for review now.

jeherve commented on projects/plugins/jetpack/extensions/blocks/goodreads/editor.js:
```diff
@@ -0,0 +1,4 @@
+import registerJetpackBlock from '../../shared/register-jetpack-block';
+import { name, settings } from '.';
+
+registerJetpackBlock( name, settings );
```
We've been working on moving away from this, and instead relying on a `block.json` to handle registration. You could check any of the other Beta blocks for some examples.
This was done in #32815.

jeherve commented on projects/plugins/jetpack/extensions/blocks/goodreads/goodreads.php:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Goodreads Block.
+ *
+ * @since 1.0.0 // Replace with the correct version number.
```
You can use `$$next-version$$` as a placeholder here, and everywhere where you need to provide a version. It will be automatically replaced during the release process.

```suggestion
 * @since $$next-version$$
```

jeherve commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-goodreads.php:
```diff
@@ -0,0 +1,65 @@
+    public function register_routes() {
+        register_rest_route(
+            'wpcom/v2',
+            '/getGoodreadsUserId',
```
I would recommend sticking to lowercase in what will be a URL. This route could be `/goodreads/user-id` for example (no need for the `get` keyword since we specify how the route can be called later).

jeherve commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-goodreads.php:
```diff
@@ -0,0 +1,65 @@
+     */
+    public function get_goodreads_user_id( $request ) {
+        $profile_id = $request->get_param( 'id' );
+        $response   = wp_remote_get( 'https://www.goodreads.com/author/show/' . $profile_id );
```
I think we can do a few things here:

1. We can build the URL separately, and escape it with `esc_url_raw` when we pass it to `wp_remote_get`.
2. We can validate the `id` param as part of the route registration, with `validate_callback`.

jeherve commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-goodreads.php:
PHPCS prefers tabs over spaces, so you would need to update this file accordingly.

jeherve commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-goodreads.php:
```diff
@@ -0,0 +1,65 @@
+        }
+
+        $body    = wp_remote_retrieve_body( $response );
+        $pattern = '/goodreads\.com\/user\/updates_rss\/(\d+)/';
```
In this pattern, is it possible to define anchors for the beginning and end of line, and is it possible to be more specific with the domain? Basically we would want to avoid matching on `https://myevildomain.com/goodreads.com/user/updates_rss/12345`, for example.

Aurorum replied:
I don't think so because the API reads the link from the middle of the Goodreads HTML (and for that reason, we also don't need to worry about matching). That being said, I've made those changes elsewhere for handling the user's inputted link.

jeherve commented on projects/plugins/jetpack/extensions/blocks/goodreads/edit.js:
```diff
@@ -0,0 +1,439 @@
+import gridExample from './grid-example.png';
+import icon from './icon';
+
+export class GoodreadsEdit extends Component {
```
In this repo we tend to prefer to implement all new components as [function components](https://react.dev/learn/your-first-component), using [hooks](https://react.dev/reference/react) to manage component state and lifecycle. Could you make the switch?

Aurorum replied:
Sorry @jeherve, I've tried pretty much everything to make this switch but I keep getting loads of console errors - feels like every time I fix one, another appears 🙃 Would you have any advice/do you have any examples or documentation in this repo?

I based the block of the Eventbrite embed which is why it's like this currently! Do I need to rewrite the block entirely?

Aurorum replied:
Ah, never mind - found some ideas from the Calendly block! Never tried it before, so let me know if there are any problems, but I've rewritten it now.

jeherve reviewed (changes requested):
Thanks for tackling this! I left a few remarks below:

********

The Grid view doesn't seem to display a grid in the editor, nor on the frontend:

<img width="802" alt="Screenshot 2023-10-10 at 17 14 26" src="https://github.com/Automattic/jetpack/assets/426388/cb06b7fa-5ade-4db8-b3af-22e839c566f0">

********

Those previews here are quite big, I think we would benefit from buttons instead, to reclaim that sidebar space:
<img width="303" alt="image" src="https://github.com/Automattic/jetpack/assets/426388/7dad40fd-72f4-4e1b-bd0d-8f77e5a921be">

********

> I'm a little stuck on how to get the block to display when embedded.

I see you're on the right track with those tests, apart from some formatting issues with the fixtures! Let's run those tests after the updates above, see how things go!

Aurorum committed:
Address feedback

Aurorum committed:
Update class-wpcom-rest-api-v2-endpoint-goodreads.php

Aurorum committed:
Update goodreads.php

Aurorum committed:
Rewrite to functional component

Aurorum commented:
Thanks for the review @jeherve! I've addressed the points you've raised. 

> The Grid view doesn't seem to display a grid in the editor, nor on the frontend

Is the problem the fact that the images aren't aligned? I've taken the styles directly from Goodreads, and "Grid" is the terminology which they use for it, but we could tidy up the styling on our end if that's something which you think would be beneficial. Alternatively, we could rename it. 

I mainly see this style as being something useful for sidebars where you might want to showcase a range of books without taking up much space. In such cases, "Grid" is probably just about a fitting description. 

> Those previews here are quite big, I think we would benefit from buttons instead, to reclaim that sidebar space

Ah, I wanted to ask about that. I agree that the previews are far too big, but I still feel like the `BlockStylesSelector` component is appropriate here since it's also an embed. Instead, I wonder if the component is broken? I remember it used to look a lot more aesthetically pleasing. 

For example, on the Eventbrite block, it now looks like this. 

<img width="278" alt="Screenshot 2023-10-11 at 21 27 26" src="https://github.com/Automattic/jetpack/assets/43215253/901d8694-8222-4717-9009-0a2c8330d76f">

In #14528, I can see that it used to look like this. Was changing this an intentional design choice or the result of a bug?

<img width="272" alt="Screenshot 2023-10-11 at 21 30 01" src="https://github.com/Automattic/jetpack/assets/43215253/153a0027-5659-47e8-a27c-4526f9f266c8">

I have a fondness to the second design, and I think that it'd make this block look much nicer too. It'd also be clearer which image applies to which style. I don't mind changing this to buttons though, and I definitely agree it'd be preferable to what it is now.

I just think that buttons would also add a layer of complexity for those trying to experiment with the two styles, and something like the old `BlockStylesSelector` design would be preferable and neat.

jeherve commented on projects/plugins/jetpack/extensions/blocks/goodreads/index.js:
What's nice about the `block.json` implementation is that we do not need that file anymore. You can check our different Beta blocks for some examples.
You can entirely remove the file.

Aurorum replied:
Ah, nice - all done! Out of interest, do I not need to worry about the strings being translatable?

jeherve commented on projects/plugins/jetpack/extensions/blocks/goodreads/goodreads.php:
Still on the topic of registering the block via `block.json`, you can check #32815 to see the few changes you would need to make to this file to full rely on `block.json`.

jeherve commented on projects/plugins/jetpack/extensions/blocks/goodreads/block.json:
```diff
@@ -0,0 +1,85 @@
+			"type": "number"
+		}
+	},
+	"editorScript": "file:../editor.js"
```
You can remove this. When we register the block via `block.json`, we do not have to also declare this since we still enqueue the JavaScript on the page via the existing big JavaScript bundle of Jetpack blocks.

jeherve commented on projects/plugins/jetpack/extensions/blocks/goodreads/edit.js:
```diff
@@ -0,0 +1,432 @@
+import React, { useState, useEffect, useRef } from 'react';
```
You shouldn't have to import `React` at all here. `useState`, `useEffect`, `useRef` can be imported from a @wordpress package instead (`@wordpress/element`)

jeherve commented on projects/plugins/jetpack/extensions/blocks/goodreads/edit.js:
```diff
@@ -0,0 +1,432 @@
+
+	useEffect( () => {
+		setUrl( props.attributes.userInput );
+	}, [ props.attributes.userInput ] );
```
Do not hesitate to [deconstruct](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) `props` at the top of your function. It will make it easier to refer to the different constants afterwards. Here is an example:
https://github.com/Automattic/jetpack/blob/260c45af538dab84b7ccdbad52b869c4c0911efb/projects/plugins/jetpack/extensions/blocks/calendly/edit.js#L19-L20

jeherve commented on projects/plugins/jetpack/extensions/blocks/goodreads/edit.js:
I think you could extract the logic that extracts data for the embed into your own [custom hook](https://react.dev/learn/reusing-logic-with-custom-hooks#extracting-your-own-custom-hook-from-a-component) in a new file. It would make the whole thing more readable and easy to understand.

You can find a number of examples in the repo, with file names that start with `use-`, if that's helpful!

Aurorum replied:
I kept getting a Lint error with this, but I've refactored this a little to make it more readable - let me know what you think!

Aurorum replied:
Done!

jeherve reviewed (commented):
Nice work so far!

I ran the different GitHub checks we have for this PR, so you now have some additional feedback about your work. I also left a few notes below.

> Is the problem the fact that the images aren't aligned? I've taken the styles directly from Goodreads, and "Grid" is the terminology which they use for it, but we could tidy up the styling on our end if that's something which you think would be beneficial.

Yes, maybe we could benefit from aligning the images so it looks more like a grid?

> I agree that the previews are far too big, but I still feel like the BlockStylesSelector component is appropriate here since it's also an embed. Instead, I wonder if the component is broken? I remember it used to look a lot more aesthetically pleasing.

You're right, it does look broken. I believe it got messed up during a recent Gutenberg update.

To solve this issue for now, I wonder if there would be a way to ditch our custom `BlockStylesSelector`, and instead use Gutenberg's own `registerBlockStyle`:
https://developer.wordpress.org/block-editor/reference-guides/block-api/block-styles/

Aurorum committed:
Address feedback

Aurorum committed:
Update edit.js

Aurorum committed:
Fix block registration

Aurorum commented:
Thanks @jeherve!

> Yes, maybe we could benefit from aligning the images so it looks more like a grid?

Done. :)

> To solve this issue for now, I wonder if there would be a way to ditch our custom BlockStylesSelector, and instead use Gutenberg's own registerBlockStyle

This is more complicated, I'm afraid. From what I understand, the reason why `BlockStylesSelector` was introduced related to the fact that `registerBlockStyle` doesn't allow for changing attributes - it basically just adds another className to the div. That's fine for a lot of blocks, but not embeds like this one. 

@scruffian - am I understanding this correctly? Basing my thoughts on this comment, which is consistent with my testing: https://github.com/Automattic/jetpack/pull/14383#issuecomment-576693765

I think that a decision would need to be made about `BlockStylesSelector` because, ideally, there should probably be consistency between the Goodreads, Calendly and Eventbrite blocks. There might be some more options, but I think the choices broadly are:

- Adding the styles from WordPress/gutenberg#16873 (which were subsequently removed in favour of buttons) to Jetpack
- Refactoring `BlockStylesSelector` to use buttons and mimic Gutenberg's behaviour; I say 'mimic', since we can't just use Gutenberg because the embed blocks require changing the style attribute. 
- Adding a more simple dropdown Setting to allow switching between the styles like any other attribute

Aurorum committed:
Fix block registration

Aurorum committed:
Merge remote-tracking branch 'upstream/trunk' into add/goodreads-block

Aurorum committed:
Refactor block

Aurorum committed:
Remove legacy widget

Aurorum commented:
Hey @jeherve, just looping back to this old one. If you have some time, I've heavily re-written this block. It now: 

- Places all the logic inside a hook so that it's easier to read
- Separates the controls from the rest of the block, also for readability
- Includes proper tests for the settings of the block (rather than just making sure anything renders)
- Removes the legacy widget 

I still can't register the two options as an official style because of the fact that it needs to be treated like an attribute. But I've resolved this by adding the icons to the Toolbar, like the Related Posts block does.

<img width="1129" alt="Screenshot 2023-12-04 at 13 15 33" src="https://github.com/Automattic/jetpack/assets/43215253/b67fece9-f8f6-4f47-b8d2-bb16cbb5b60c">

Overall, I'm hoping that this works and looks better - both within the code and within the Editor. :) Thanks!!

Aurorum committed:
Indentation fix

Aurorum committed:
Merge remote-tracking branch 'upstream/trunk' into add/goodreads-block

scruffian commented:
> @scruffian - am I understanding this correctly? Basing my thoughts on this comment, which is consistent with my testing: https://github.com/Automattic/jetpack/pull/14383#issuecomment-576693765

That sounds right but it's been a long time since I worked on it!

Aurorum committed:
Merge remote-tracking branch 'upstream/trunk' into add/goodreads-block

Aurorum committed:
Fix tests

Aurorum committed:
Manually update files

Aurorum committed:
Manually update file

Aurorum commented:
Tests should be fixed now

Aurorum committed:
Merge branch 'trunk' into add/goodreads-block

Aurorum commented:
Hey @jeherve - just looping back to this old one. I can see that the tests are passing except for the WordPress.com Tests. However, I don’t seem able to access the logs and see why they’re failing.

Are there any changes needed from my end?

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/goodreads-block

jeherve commented:
@Aurorum I think we're getting there! Thanks for working on that! Everything works well for me, except for the number of items. Does this work for you? When I refresh the editor, the number resets back to 5.

Aurorum committed:
Fix bookNumber not saving

Aurorum commented:
Whoops, that was just me registering the wrong type - not sure how I missed that! Should be fixed now. :) 

Thanks for taking another look!

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/goodreads-block

jeherve reviewed (approved):
This looks good. Let's merge that first version!

jeherve commented:
Related diff: D138696-code
